### PR TITLE
ECS Service and cluster scaling

### PR DESCRIPTION
**Pull request by ThomasVdBerge**
## Description
Added module to simplify the adding of ECS service/cluster scaling.

## Related Issue

https://github.com/skyscrapers/diekeure/issues/65

## Motivation and Context
```
+ module.service-scaling.aws_appautoscaling_policy.scale_down
    adjustment_type:                                        "ChangeInCapacity"
    arn:                                                    "<computed>"
    cooldown:                                               "60"
    metric_aggregation_type:                                "Maximum"
    name:                                                   "default-test-service-scale-in"
    policy_type:                                            "StepScaling"
    resource_id:                                            "service/test/test"
    scalable_dimension:                                     "ecs:service:DesiredCount"
    service_namespace:                                      "ecs"
    step_adjustment.#:                                      "1"
    step_adjustment.2173517692.metric_interval_lower_bound: ""
    step_adjustment.2173517692.metric_interval_upper_bound: "0"
    step_adjustment.2173517692.scaling_adjustment:          "-1"

+ module.service-scaling.aws_appautoscaling_policy.scale_up
    adjustment_type:                                        "ChangeInCapacity"
    arn:                                                    "<computed>"
    cooldown:                                               "60"
    metric_aggregation_type:                                "Maximum"
    name:                                                   "default-test-service-scale-out"
    policy_type:                                            "StepScaling"
    resource_id:                                            "service/test/test"
    scalable_dimension:                                     "ecs:service:DesiredCount"
    service_namespace:                                      "ecs"
    step_adjustment.#:                                      "1"
    step_adjustment.3459629261.metric_interval_lower_bound: ""
    step_adjustment.3459629261.metric_interval_upper_bound: "0"
    step_adjustment.3459629261.scaling_adjustment:          "1"

+ module.service-scaling.aws_appautoscaling_target.ecs_target
    max_capacity:       "10"
    min_capacity:       "2"
    resource_id:        "service/test/test"
    role_arn:           "${aws_iam_role.ecs-autoscale-role.arn}"
    scalable_dimension: "ecs:service:DesiredCount"
    service_namespace:  "ecs"

+ module.service-scaling.aws_autoscaling_policy.ecs_cluster_scale_in
    adjustment_type:         "PercentChangeInCapacity"
    arn:                     "<computed>"
    autoscaling_group_name:  "ecs-cluster-test"
    cooldown:                "300"
    metric_aggregation_type: "<computed>"
    name:                    "default-test-cluster-scale-in"
    policy_type:             "SimpleScaling"
    scaling_adjustment:      "-50"

+ module.service-scaling.aws_autoscaling_policy.ecs_cluster_scale_out
    adjustment_type:         "PercentChangeInCapacity"
    arn:                     "<computed>"
    autoscaling_group_name:  "ecs-cluster-test"
    cooldown:                "300"
    metric_aggregation_type: "<computed>"
    name:                    "default-test-cluster-scale-out"
    policy_type:             "SimpleScaling"
    scaling_adjustment:      "100"

+ module.service-scaling.aws_cloudwatch_metric_alarm.ecs_cluster_scale_in_alarm
    actions_enabled:                       "true"
    alarm_actions.#:                       "<computed>"
    alarm_description:                     "This metric monitor ecs CPU utilization down"
    alarm_name:                            "default-test-ECSClusterScaleInAlarm"
    comparison_operator:                   "LessThanOrEqualToThreshold"
    dimensions.%:                          "1"
    dimensions.ClusterName:                "test"
    evaluate_low_sample_count_percentiles: "<computed>"
    evaluation_periods:                    "4"
    metric_name:                           "CPUReservation"
    namespace:                             "AWS/ECS"
    period:                                "120"
    statistic:                             "Average"
    threshold:                             "25"
    treat_missing_data:                    "missing"

+ module.service-scaling.aws_cloudwatch_metric_alarm.ecs_cluster_scale_out_alarm
    actions_enabled:                       "true"
    alarm_actions.#:                       "<computed>"
    alarm_description:                     "This metric monitor ecs CPU utilization up"
    alarm_name:                            "default-test-ECSClusterScaleOutAlarm"
    comparison_operator:                   "GreaterThanOrEqualToThreshold"
    dimensions.%:                          "1"
    dimensions.ClusterName:                "test"
    evaluate_low_sample_count_percentiles: "<computed>"
    evaluation_periods:                    "4"
    metric_name:                           "CPUReservation"
    namespace:                             "AWS/ECS"
    period:                                "120"
    statistic:                             "Average"
    threshold:                             "75"
    treat_missing_data:                    "missing"

+ module.service-scaling.aws_cloudwatch_metric_alarm.ecs_service_scale_in_alarm
    actions_enabled:                       "true"
    alarm_actions.#:                       "<computed>"
    alarm_description:                     "This metric monitor ecs CPU utilization down"
    alarm_name:                            "default-test-ECSServiceScaleInAlarm"
    comparison_operator:                   "LessThanOrEqualToThreshold"
    dimensions.%:                          "2"
    dimensions.ClusterName:                "test"
    dimensions.ServiceName:                "test"
    evaluate_low_sample_count_percentiles: "<computed>"
    evaluation_periods:                    "4"
    metric_name:                           "CPUUtilization"
    namespace:                             "AWS/ECS"
    period:                                "120"
    statistic:                             "Average"
    threshold:                             "25"
    treat_missing_data:                    "missing"

+ module.service-scaling.aws_cloudwatch_metric_alarm.ecs_service_scale_out_alarm
    actions_enabled:                       "true"
    alarm_actions.#:                       "<computed>"
    alarm_description:                     "This metric monitor ecs CPU utilization up"
    alarm_name:                            "default-test-ECSServiceScaleOutAlarm"
    comparison_operator:                   "GreaterThanOrEqualToThreshold"
    dimensions.%:                          "2"
    dimensions.ClusterName:                "test"
    dimensions.ServiceName:                "test"
    evaluate_low_sample_count_percentiles: "<computed>"
    evaluation_periods:                    "4"
    metric_name:                           "CPUUtilization"
    namespace:                             "AWS/ECS"
    period:                                "120"
    statistic:                             "Average"
    threshold:                             "75"
    treat_missing_data:                    "missing"

+ module.service-scaling.aws_iam_role.ecs-autoscale-role
    arn:                "<computed>"
    assume_role_policy: "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Action\": \"sts:AssumeRole\",\n      \"Principal\": {\n        \"Service\": \"application-autoscaling.amazonaws.com\"\n      },\n      \"Effect\": \"Allow\"\n    }\n  ]\n}\n"
    create_date:        "<computed>"
    name:               "ecs-scale-default-test"
    path:               "/"
    unique_id:          "<computed>"

+ module.service-scaling.aws_iam_role_policy_attachment.ecs_autoscale
    policy_arn: "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceAutoscaleRole"
    role:       "${aws_iam_role.ecs-autoscale-role.id}"

+ module.service-scaling.aws_iam_role_policy_attachment.ecs_cloudwatch
    policy_arn: "arn:aws:iam::aws:policy/CloudWatchFullAccess"
    role:       "${aws_iam_role.ecs-autoscale-role.id}"


Plan: 25 to add, 0 to change, 0 to destroy.
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: scaling/main.tf [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  }
+  alarm_description = "This metric monitor ecs CPU utilization down"
+  alarm_actions     = ["${aws_autoscaling_policy.ecs_cluster_scale_in.arn}"]
+}
+
+resource "aws_autoscaling_policy" "ecs_cluster_scale_in" {
+  name                   = "${terraform.env}-${var.cluster_name}-cluster-scale-in"
+  policy_type            = "SimpleScaling"
+  adjustment_type        = "PercentChangeInCapacity"
+  scaling_adjustment     = -50
```
Comment (venturel) [inline scaling/main.tf]: Would it be possible to create a var with default vallue for this?

File: scaling/main.tf [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  cooldown               = 300
+  autoscaling_group_name = "${var.ecs_autoscale_group_name}"
+}
+
+resource "aws_autoscaling_policy" "ecs_cluster_scale_out" {
+  name                   = "${terraform.env}-${var.cluster_name}-cluster-scale-out"
+  policy_type            = "SimpleScaling"
+  adjustment_type        = "PercentChangeInCapacity"
+  scaling_adjustment     = 100
```
Comment (venturel) [inline scaling/main.tf]: also for this

File: scaling/main.tf [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+resource "aws_cloudwatch_metric_alarm" "ecs_service_scale_out_alarm" {
+  alarm_name          = "${terraform.env}-${var.cluster_name}-ECSServiceScaleOutAlarm"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "${var.evaluation_periods}"
+  metric_name         = "CPUUtilization"
+  namespace           = "AWS/ECS"
+  period              = "${var.period_down}"
+  statistic           = "${var.statistic}"
+  threshold           = "75"
```
Comment (venturel) [inline scaling/main.tf]: also for this

File: scaling/main.tf [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+resource "aws_cloudwatch_metric_alarm" "ecs_service_scale_in_alarm" {
+  alarm_name          = "${terraform.env}-${var.cluster_name}-ECSServiceScaleInAlarm"
+  comparison_operator = "LessThanOrEqualToThreshold"
+  evaluation_periods  = "${var.evaluation_periods}"
+  metric_name         = "CPUUtilization"
+  namespace           = "AWS/ECS"
+  period              = "${var.period_down}"
+  statistic           = "${var.statistic}"
+  threshold           = "25"
```
Comment (venturel) [inline scaling/main.tf]: and this

File: scaling/main.tf [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  cooldown                = 60
+  metric_aggregation_type = "Maximum"
+  name                    = "${terraform.env}-${var.cluster_name}-service-scale-in"
+  resource_id             = "service/${var.cluster_name}/${var.service_name}"
+  scalable_dimension      = "ecs:service:DesiredCount"
+  service_namespace       = "ecs"
+
+  step_adjustment {
+    metric_interval_upper_bound = "${var.lowerbound}"
+    scaling_adjustment          = "-1"
```
Comment (venturel) [inline scaling/main.tf]: and this

File: scaling/main.tf [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  cooldown                = 60
+  metric_aggregation_type = "Maximum"
+  name                    = "${terraform.env}-${var.cluster_name}-service-scale-out"
+  resource_id             = "service/${var.cluster_name}/${var.service_name}"
+  scalable_dimension      = "ecs:service:DesiredCount"
+  service_namespace       = "ecs"
+
+  step_adjustment {
+    metric_interval_upper_bound = "${var.upperbound}"
+    scaling_adjustment          = "1"
```
Comment (venturel) [inline scaling/main.tf]: and this

File: README.md [hunk 7]
```diff
@@ -22,3 +22,31 @@ module "ecs_cluster" {
   environment = "${terraform.env}"
 }
 ```
+
+## scaling
+Setup a ECS cluster and service scaling.
+
+### Available variables:
+ * [`cluster_name`]: String(required): The name of the ECS cluster.
+ * [`service_name`]: String(required): The name of the ECS cluster.
```
Comment (venturel) [inline README.md]: not sure this is the right description

File: scaling/main.tf [hunk 8]
```diff
@@ -0,0 +1,126 @@
+resource "aws_cloudwatch_metric_alarm" "ecs_cluster_scale_out_alarm" {
+  alarm_name          = "${terraform.env}-${var.cluster_name}-ECSClusterScaleOutAlarm"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "${var.evaluation_periods}"
+  metric_name         = "CPUReservation"
+  namespace           = "AWS/ECS"
+  period              = "${var.period_down}"
+  statistic           = "${var.statistic}"
+  threshold           = "75"
```
Comment (venturel) [inline scaling/main.tf]: Would it be possible to create a var with default vallue for this?

File: scaling/main.tf [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+resource "aws_cloudwatch_metric_alarm" "ecs_cluster_scale_in_alarm" {
+  alarm_name          = "${terraform.env}-${var.cluster_name}-ECSClusterScaleInAlarm"
+  comparison_operator = "LessThanOrEqualToThreshold"
+  evaluation_periods  = "${var.evaluation_periods}"
+  metric_name         = "CPUReservation"
+  namespace           = "AWS/ECS"
+  period              = "${var.period_down}"
+  statistic           = "${var.statistic}"
+  threshold           = "25"
```
Comment (venturel) [inline scaling/main.tf]: also here

File: scaling/variables.tf [hunk 10]
```diff
@@ -0,0 +1,44 @@
+variable "cluster_name" {
+  description = "Name of the cluster"
+}
+
+variable "service_name" {
+  description = "Name of the service"
+}
+
+variable "evaluation_periods" {
+  description = "Name of the service"
```
Comment (venturel) [inline scaling/variables.tf]: I'd change or remove the description

**Review by venturel — CHANGES_REQUESTED** [review]

**Comment by venturel** [comment]
Also @ThomasVdBerge why don't you use the scale module for the instance autoscaling? And you call scale in/out and scale up/down. Would it be possible to use just one of them? Personally I prefer up/down.

File: scaling/iam.tf [hunk 11]
```diff
@@ -0,0 +1,27 @@
+resource "aws_iam_role" "ecs-autoscale-role" {
+  name = "ecs-scale-${terraform.env}-${var.cluster_name}"
```
Comment (ringods) [inline scaling/iam.tf]: If this module is aimed at configuring the service auto scaling, do not create cluster specific resources in this module. Trying to call this module to configure the service scaling for two services on the same cluster will lead to an error when calling it the second time as the name is already taken. The cluster resources should be created in the module setting up the ECS cluster.

**Review by ringods — CHANGES_REQUESTED** [review]

File: scaling/iam.tf (hunk 11 above)
Comment (ThomasVdBerge) [reply scaling/iam.tf]: You're right, the var cluster should not be the only one in the name. Will also add service var. (Only adding service var would give different issues)

Comment (ringods) [reply scaling/iam.tf]: @ThomasVdBerge you didn't fix the naming of the iam resources with your last commit, so my initial remark is still valid. ;-)

**Review by ringods — APPROVED** [review]

**Review by venturel — APPROVED** [review]

**Pull request merged**